src/main/java/com/faforever/client/fx/PlayerRatingChart.java

```diff
@@ -0,0 +1,187 @@
+  }
+
+  public void initializeTooltip(UiService uiService) {
+    if (available) {
+      this.tooltipController = uiService.loadFxml("theme/chat/player_rating_chart_tooltip.fxml");
+      prepareHoverTooltip();
+    }
+  }
+
+  private void prepareHoverTooltip() {
+    hoverTooltip = JavaFxUtil.createCustomTooltip(tooltipController.getRoot());
+    hoverTooltip.setAutoFix(false);
+    hoverTooltip.setHideDelay(Duration.ZERO);
+    hoverTooltip.setShowDelay(Duration.ZERO);
+    hoverTooltip.setShowDuration(Duration.INDEFINITE);
+    Tooltip.install(chartBackground, hoverTooltip);
+  }
+
+  private void setValues(MouseEvent event) {
+    if (valid) {
+      double x = event.getX();
+      long dateValueInSec = getDisplayedDateValue(x);
+      Double rating = ratingMap.get((int) x);
+      if (rating != null && dateValueInSec != Long.MIN_VALUE) {
+        tooltipController.setDateAndRating(dateValueInSec, rating.intValue());
+      } else {
+        tooltipController.clear();
+      }
+    }
+  }
+
+  private void moveTooltipAndLineToCursor(MouseEvent event) {
+    hoverTooltip.setX(event.getScreenX() + OFFSET_TOOLTIP_FROM_CURSOR_BY_X);
+    hoverTooltip.setY(event.getScreenY() + OFFSET_TOOLTIP_FROM_CURSOR_BY_Y);
+    verticalLine.setLayoutX(event.getX());
+  }
+
+  private void recalculateData() {
+    ratingMap.clear();
+    for (int i = 0; i < getData().size(); i++) {
+      Series<Double, Double> series = getData().get(i);
+      if (series.getNode() instanceof Path) {
+        buildData(((Path) series.getNode()).getElements());
+        break;
+      }
+    }
+  }
+
+  private void buildData(List<PathElement> elements) {
+    List<LineTo> lines = elements.stream().filter((e) -> e instanceof LineTo).map((e) -> (LineTo) e).collect(Collectors.toList());
+    if (lines.size() < 2) {
+      return;
+    }
+
+    for (int i = 0; i < lines.size() - 1; i++) {
+      LineTo line1 = lines.get(i);
+      LineTo line2 = lines.get(i + 1);
+      int x1 = (int) Math.floor(line1.getX());
+      int x2 = (int) Math.floor(line2.getX());
+      int rating1 = getDisplayedRatingValue(line1.getY());
+      int rating2 = getDisplayedRatingValue(line2.getY());
+      if (rating1 == Integer.MIN_VALUE || rating2 == Integer.MIN_VALUE) {
+        continue;
+      }
+      int distance = x2 - x1;
+      if (distance > 1) {
+        double k = (double) (rating2 - rating1) / distance;
+        double value = rating1;
+        while (x1 < x2) {
+          putMapRatingValue(++x1, value += k);
+        }
+      } else {
+        putMapRatingValue(x1, rating1);
+      }
+    }
+  }
+
+  private void putMapRatingValue(int x, double rating) {
+    Double currentRating = ratingMap.get(x);
+    ratingMap.put(x, currentRating != null ? (currentRating + rating) / 2 : rating);
+  }
+
+  private long getDisplayedDateValue(double displayPosition) {
+    Double value = getXAxis().getValueForDisplay(displayPosition);
+    return value != null ? value.longValue() : Long.MIN_VALUE;
```

Comment:
```
return Optional.ofNullable(getXAxis().getValueForDisplay(displayPosition))
             .map(ClassNameGoesHere::longValue)
             .else(Long.MIN_VALUE)
```
Just a suggestion, this temp value variable and tertiary op look very verbose, not sure if it's any better though.

Reply:
Done. :)